In [32]:
import json
from collections import Counter

import pandas as pd

# Task 1

In [2]:
likes_df = pd.read_csv('train_likes.csv')

In [3]:
likes_df.head()

user_id                           item_id  \
0  612d8e8eef05acff3278c061ec10f704  7aa5d00445cb9d61d1739dd0df9a0a88   
1  71a7f1d1be96603971ba66e4a17e845c  5edaf734b432e5cc954a10b59cb97e70   
2  6eaa117728d50265e6b2ac24a80e04ae  8ad97d075fce19c2d182eb2a4539aa1c   
3  5d9db6ab742755197343505bccfad516  aa5f2ca699da42e467e550f9f071fb3f   
4  3000a163610654f1fa181e74136d2d35  8142d0e687c1c7a317ed9673db9f11a4   

                            channel          time  
0  1f0e3dad99908345f7439f8ffabdffc4  1.389734e+09  
1  ec5decca5ed3d6b8079e2e7e7bacc9f2  1.390459e+09  
2  98f13708210194c475687be6106a3b84  1.391064e+09  
3  98f13708210194c475687be6106a3b84  1.391053e+09  
4  c74d97b01eae257e44aa9d5bade97baf  1.392381e+09

In [4]:
likes_df.channel.value_counts().mean()

106.05433746425167

In [5]:
(likes_df.item_id.value_counts() > 5).value_counts()

False    20200
True      3691
Name: item_id, dtype: int64

### Ans

106.05433746425167

3691

# Task 2

In [52]:
items_data = dict()
for item in json.load(open('items.json')):
    items_data[item['id']] = item

In [109]:
likes_for_genres = Counter()
likes_for_no_genres = 0

for item_id, likes in likes_df.item_id.value_counts().iteritems():
    data = items_data.get(item_id, None)
    if data is not None and 'genre' in data:
        likes_for_genres[data['genre']] += likes
    else:
        likes_for_no_genres += likes

-----

In [112]:
likes_df.channel.value_counts()[:10]

1f0e3dad99908345f7439f8ffabdffc4    8348
c74d97b01eae257e44aa9d5bade97baf    8057
98f13708210194c475687be6106a3b84    6079
1ff1de774005f8da13f42943881c655f    5853
70efdf2ec9b086079795c442636b55fb    4619
c0c7c76d30bd3dcaefc96f40275bdc0a    3257
f90f2aca5c640289d0a29417bcb63a37    3037
6f4922f45568161a8cdf4ad2299f6d23    2925
4e732ced3463d06de0ca9a15b6153677    2528
5c936263f3428a40227908d5a3847c0b    2483
Name: channel, dtype: int64

In [113]:
likes_for_genres_by_channels = []
likes_for_no_genres_by_channels = []

for channel, channel_likes in likes_df.channel.value_counts()[:10].iteritems():
    likes_for_genres_by_channel = Counter()
    likes_for_no_genres_by_channel = 0
    
    for item_id, likes in likes_df[likes_df.channel == channel].item_id.value_counts().iteritems():
        data = items_data.get(item_id, None)
        if data is not None and 'genre' in data:
            likes_for_genres_by_channel[data['genre']] += likes
        else:
            likes_for_no_genres_by_channel += likes
            
    likes_for_genres_by_channels.append((channel, likes_for_genres_by_channel))
    likes_for_no_genres_by_channels.append((channel, likes_for_no_genres_by_channel))

-----

In [135]:
for i in range(9):
    print(likes_for_genres[i], end=' ')

14007 28151 14237 5231 2494 6543 4383 1167 27 

In [144]:
for i, it in enumerate(likes_for_genres_by_channels):
    channel, genres_counter = it
    print(channel, end=' ')
    for j in range(9):
        print(genres_counter[j], end=' ')
    print(likes_for_no_genres_by_channels[i][1], end='')
    print()
    print()

1f0e3dad99908345f7439f8ffabdffc4 1138 2913 1750 0 13 0 487 0 0 2047

c74d97b01eae257e44aa9d5bade97baf 2364 1689 873 496 366 209 13 0 0 2047

98f13708210194c475687be6106a3b84 1882 1256 1806 3 6 0 46 0 1 1079

1ff1de774005f8da13f42943881c655f 73 157 23 111 70 3500 0 0 0 1919

70efdf2ec9b086079795c442636b55fb 684 99 494 332 213 29 0 0 0 2768

c0c7c76d30bd3dcaefc96f40275bdc0a 50 391 315 0 12 0 1095 0 0 1394

f90f2aca5c640289d0a29417bcb63a37 507 281 76 71 0 0 13 0 21 2068

6f4922f45568161a8cdf4ad2299f6d23 334 380 472 200 257 4 29 0 0 1249

4e732ced3463d06de0ca9a15b6153677 31 1198 282 124 13 0 3 0 0 877

5c936263f3428a40227908d5a3847c0b 1187 546 380 4 0 0 6 0 0 360



### Ans

14007 28151 14237 5231 2494 6543 4383 1167 27 35011

-----

1f0e3dad99908345f7439f8ffabdffc4 1138 2913 1750 0 13 0 487 0 0 2047


c74d97b01eae257e44aa9d5bade97baf 2364 1689 873 496 366 209 13 0 0 2047


98f13708210194c475687be6106a3b84 1882 1256 1806 3 6 0 46 0 1 1079


1ff1de774005f8da13f42943881c655f 73 157 23 111 70 3500 0 0 0 1919


70efdf2ec9b086079795c442636b55fb 684 99 494 332 213 29 0 0 0 2768


c0c7c76d30bd3dcaefc96f40275bdc0a 50 391 315 0 12 0 1095 0 0 1394


f90f2aca5c640289d0a29417bcb63a37 507 281 76 71 0 0 13 0 21 2068


6f4922f45568161a8cdf4ad2299f6d23 334 380 472 200 257 4 29 0 0 1249


4e732ced3463d06de0ca9a15b6153677 31 1198 282 124 13 0 3 0 0 877


5c936263f3428a40227908d5a3847c0b 1187 546 380 4 0 0 6 0 0 360
